drivers: counter: Add detection of late alarm request for nRF RTC and TIMER drivers

nordic-krch opened the pull request:
Alarm can be set too late due to procedure interruption or
miscalculation or by RTC limitation (setting now + 1 alarm
won't trigger an event). Added an algorithm to detect late
setting. If such situation is detected user callback is
called from the caller context.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: counter: nrf_rtc: Add detection of late alarm request

Implemented latest extensions to the counter API related to
detection of alarms being set too late and short relative alarms.
Implementation could not be realized on nrfx_rtc driver thus
driver has been reimplemented based on nrf_rtc hal.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

pizi-nordic commented on drivers/counter/Kconfig.nrfx:
```diff
@@ -112,6 +112,19 @@ config COUNTER_RTC0_PRESCALER
+	  Guard period should be set to the maximal latency that can occur when
+	  controlling the counter (e.g. for how long operation may be
+	  interrupted). Guard period limits maximal value that can be set to
+	  alarm to top value - guard period in ticks. Longer alarms will be
```
... in counter ticks ...

pizi-nordic replied:
... in counter ticks ...

We have to avoid confusion with system ticks.

pizi-nordic commented on drivers/counter/Kconfig.nrfx:
```diff
@@ -112,6 +112,19 @@ config COUNTER_RTC0_PRESCALER
+	  Guard period should be set to the maximal latency that can occur when
+	  controlling the counter (e.g. for how long operation may be
+	  interrupted). Guard period limits maximal value that can be set to
+	  alarm to top value - guard period in ticks. Longer alarms will be
```
Longer alarms .... will expire immediately?
Sounds weird...

nordic-krch replied:
i elaborated a bit more. If it is still unclear please provide your suggestion.

pizi-nordic replied:
Now it is better.

pizi-nordic commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -72,18 +76,81 @@ static u32_t counter_nrfx_read(struct device *dev)
+	{
+		LOG_INST_INF(nrfx_config->log,
+			"Delay introduced due to risk of pre-firing.");
+		NRFX_DELAY_US(33);
```
Why 33? (we need comment or define)

nordic-krch replied:
fixed

pizi-nordic commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -72,18 +76,81 @@ static u32_t counter_nrfx_read(struct device *dev)
+	if (((val - now) & top) == 1) {
+		LOG_INST_INF(nrfx_config->log,
+			"Delay introduced due to short CC.");
+		NRFX_DELAY_US(33);
```
33 used in more than one place => define.

nordic-krch replied:
fixed

pizi-nordic commented on drivers/counter/Kconfig.nrfx:
```diff
@@ -112,6 +112,19 @@ config COUNTER_RTC0_PRESCALER
 	help
 	  Frequency = 32768 / (prescaler+1).
 
+config COUNTER_RTC0_GUARD_PERIOD_MS
```
We have to include this guard in counter_api_get_max_relative_alarm.

nordic-krch replied:
done

pizi-nordic commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -13,7 +13,8 @@
 LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL);
 
-#define RTC_CLOCK 32768
+#define LF_FREQ 32768
```
Is there as system-level define for that? If no, could we add clock-frequency node to the RTC?

nordic-krch replied:
as already mentioned in another comment, i will handle DTS related stuff in separate commit. ok?

nordic-krch replied:
as #16135 went in, things have been moved to device tree.

nordic-krch commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -26,6 +35,8 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL);
 #define TOP_CH 0
 #define COUNTER_TOP_INT NRFX_RTC_INT_COMPARE0
 
+#define MS2TICKS(ms, prescaler) (((ms) * 32768/((prescaler) + 1)) / 1000)
```
fixed

carlescufi reviewed (approved):
Overall looks good, needs a second review from someone with timer knowledge

pabigot commented on dts/bindings/rtc/nordic,nrf-rtc.yaml:
```diff
@@ -23,6 +23,20 @@ properties:
+    # interpretted as set too late and will expire immediately. If instance
+    # is used only from interrupt context value probably can be decreased.
+    # However, if used from low priority thread it may need to be increased.
+    guard-period:
```
I'm not sure what this is trying to describe.

RTC requires that a value stored in a `CC` register be at least 2 ticks  past the current `COUNTER` to guarantee a `COMPARE` event fires.  Is it that `2` that's supposed to be captured here?  I'll assume so, though for some reason the value in the DTS files is `10`.

That limitation has no relation to interrupt priority, so much of the description brings in irrelevant concepts.  Also `guard-period` does not suggest what this value is for in terms related to RTC or counter.  Something like `compare-tick-offset` would be closer, though I don't love it.

Further, whatever this property describes probably applies to other SOC asynchronous timer infrastructure.  The property should be described in the parent `rtc.yaml`.  Something like it probably needs to be in a parent `timer.yaml` that does not currently exist.

The description field should be useful, like the other ones in the parent `rtc.yaml`.

nordic-krch replied:
Value that should be set to guard-period is the maximal potential preemption period that can occur during setting the alarm. If alarm is only controlled from interrupt context then guard period can be small, if from any context then it must be increased. Increasing guard-period reduced maximal alarm value that can be set (e.g in RTC case 2^24-guard-period-in-ticks) so if user is unsure it should be increased. By default i set it to 10ms which i thought should be ok for rtos system.

I agree that it's difficult to describe it well. I will try to look into improving this description.

pabigot replied:
No.

The "maximal potential preemption period" is necessarily dependent on what the application does.  Maybe I've got an app that has a critical data collection that takes 100 ms and cannot be interrupted.

This is not a hardware-related property, and does not belong in the device tree (unlike the 2 tick counter offset, which does).  If you have to have it to make the algorithm work, it should be a property passed to the counter when the application configures it, just as with top value.

And wherever it goes, you should specify the units of the value; I had no idea it wasn't RTC ticks.  Milliseconds for time resolution is a Zephyr design decision  unrelated to Nordic hardware; it's also one that will very soon have to be reconsidered because it's a poor choice given current MCU operating frequencies.

nordic-krch replied:
> This is not a hardware-related property, and does not belong in the device tree

I think there is an overall assumption in zephyr that driver instance configuration is part of device tree. Whether it's hw related or not. I initially was against but accepted as imo it's better to have instance configuration in one place rather than split across two configuration tools.

nordic-krch replied:
>Milliseconds for time resolution is a Zephyr design decision unrelated to Nordic hardware; it's also one that will very soon have to be reconsidered because it's a poor choice given current MCU operating frequencies.

I think that milliseconds is more user friendly since it is still rough value but I can change it to ticks if you insist and even set the default to be much bigger (like 32768). That limits RTC range to 511s instead of 512s (assuming 0 prescaler and no change in top value) which will be unnoticed in majority of cases.

I need to decrease it for the tests where i use smaller top value.

> it should be a property passed to the counter when the application configures it, just as with top value.

Would you see that as part of the API, something like: `set_top_value(dev, ... , guard)`?
@pizi-nordic @anangl opinions? @pabigot pros, cons?

pabigot replied:
> > This is not a hardware-related property, and does not belong in the device tree
> 
> I think there is an overall assumption in zephyr that driver instance configuration is part of device tree. Whether it's hw related or not. I initially was against but accepted as imo it's better to have instance configuration in one place rather than split across two configuration tools.

That is certainly not the impression I had, based on the expectation that device-tree content was to be OS-independent (shared with other projects).  @galak ?

pabigot commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -81,18 +94,97 @@ static u32_t counter_nrfx_read(struct device *dev)
+		now = nrf_rtc_counter_get(reg);
+	}
+
+	/* If value is in window spanning from short in the past to one tick
```
Why does this allow the alarm to fire before the time has been reached?  If `val-1==now` the alarm should not fire.

Or is that because in the test above you already delayed that extra tick?

nordic-krch replied:
the assumption here is that CC is set and after that it is checked if we set for next tick. Delay if yes. However, i just noticed that it may lead to prefiring if prescaler is different than 0. i need to look into that.

pabigot commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -81,18 +94,97 @@ static u32_t counter_nrfx_read(struct device *dev)
+ */
+static u32_t tick_diff(u32_t val, u32_t old, u32_t top)
+{
+	if (top & (top + 1)) {
```
Should be `top & (top - 1)`.   Try `top=7`.

pabigot replied:
Nope, this was right; I'm not used to inclusive upper bounds.

pabigot commented:
Referencing issue #12626, this solution lacks the functionality of allowing the caller to determine how late alarms are handled.  Invoking the callback immediately may be catastrophic in some applications: instead the attempt to set should fail.

pabigot commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -81,18 +94,97 @@ static u32_t counter_nrfx_read(struct device *dev)
+	prev_val = nrf_rtc_cc_get(reg, chan);
+	nrf_rtc_cc_set(reg, chan, val);
+
+	/* If prev CC setting equals or is just in front of the counter then
```
This hints at the role of the `2` that I suggested might be in the RTC device tree, so probably `guard-period` isn't what I thought it was.  If so, I don't think it belongs in device-tree because it's not related to a hardware capability.

pabigot reviewed (dismissed):
Updated status:

I have not run this, nor closely examined whether the reworked set_cc logic appears correct.

I'm currently NAKing it for these issues:
* This introduces a behavior change that relates to API, without any evidence it's been approved by non-Nordic vendors or undergone API "committee" review;
* The material that's currently in device tree is based on expected application behavior, not hardware features, so belongs in Kconfig rather than device tree.

However, I won't ever approve this because it doesn't appear to meet the functional requirements that motivated #12626 (which requires changes to the user API as well as to behavior).  I believe this is technically the wrong solution.

My NAK can be overridden, or these issues can be addressed and my review dismissed (I can't change it to "comment only" because github doesn't allow it).

pabigot commented:
My review comment that led to my selecting "Request Changes" was incorrect, so my review should be "Comment Only".  Github won't let me change it to anything but "Approve", so I'm leaving it as is--but as it says, it's not intended as a NAK.

nordic-krch commented:
@pabigot thanks for the comments. I will try to describe the issue again.

There are 2 problems there. First one is possibility of preemption which can occur at any moment for unknown period. Second is the RTC peripheral which is clocked with 32khz so when you write to any register, actual write will happen on next LFCLK tick.

Then we have potential issues like:
- previous CC being set to now (or next tick). If we write new CC and enable interrupts we may get interrupt immediately
- setting new CC that is already in the past
- setting new CC to now+1 which will not generate IRQ

I've tried to cover those scenarios and test them. Tests are not yet in PR but will come soon.

I will your comments one by one then.

nordic-krch commented:
> Referencing issue #12626, this solution lacks the functionality of allowing the caller to determine how late alarms are handled. Invoking the callback immediately may be catastrophic in some applications: for those the attempt to set should fail.

If callback is never triggered before expected timeout the only issue i see is that context of the callback is changed. Is that the issue or you see other potential failures. Can you elaborate?

My assumption was that immediate expiration simplifies handling as returned code requires user to react and reaction most likely will be to handle the timeout.

pabigot commented:
> > Referencing issue #12626, this solution lacks the functionality of allowing the caller to determine how late alarms are handled. Invoking the callback immediately may be catastrophic in some applications: for those the attempt to set should fail.
> 
> If callback is never triggered before expected timeout the only issue i see is that context of the callback is changed. Is that the issue or you see other potential failures. Can you elaborate?

There are two use-cases for handling late alarms:
* Best effort is ok: fire anyway.  The application can react to lateness within the alarm action.
* Best effort is not ok: do no harm.  The alarm action would cause harm because an application invariant has been missed and some sort of recovery is necessary.

A complete API would handle both.  Note lateness can be from late-to-callback, which isn't being addressed here, or late-to-set, which is.

Made-up example: The alarm controls a valve to fill containers spaced on a moving belt so the container is only present for 50 ms with a 10 ms gap (assume the valve is controlled by PPI to avoid mixing late-to-set with late-to-callback).  If the alarm to open is late the container won't be filled completely, which is a production error: better to leave it closed and not waste product.  If the alarm to close is late I've got spillage and need to shut down the line.  In neither case is executing the alarm action acceptable, and in both cases the application needs to be told that the alarm didn't function as desired.

nordic-krch commented:
@pabigot 
> A complete API would handle both. Note lateness can be from late-to-callback, which isn't being addressed here, or late-to-set, which is.

Both type of lateness can be discovered in the callback by comparing `ticks` which are part of the callback with `counter_read()` performed in the callback context. Wouldn't that satisfy your requirement while keeping api simpler?

anangl commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -81,18 +94,97 @@ static u32_t counter_nrfx_read(struct device *dev)
+
+	/* If prev CC setting equals or is just in front of the counter then
+	 * there is a risk that before new CC will be set after enabling event
+	 * previous CC will generate COMPARE event. In such risk delay must be
```
I'm not sure I understand the risk. I assume that it relates to the fact that the value written to the `CC` register will be actually applied in the peripheral at the next LFCLK tick. But isn't it the case also with the `EVTEN` register (I mean, isn't the application of its new value delayed in the same way)?

nordic-krch replied:
I've just recalled that `EVTEN` usage here is useless (by @pabigot observation #15624). I've been trying to stress this procedure even more today since i've seen that it may be still possible to get interrupt on previous CC. I managed to caught it once. Working on improving test and algorithm.

anangl commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -81,18 +94,97 @@ static u32_t counter_nrfx_read(struct device *dev)
+	nrf_rtc_event_enable(reg, int_mask);
+
+	/* If value is set to now + 1 wait one LFCLK cycle to ensure that it is
+	 * written to RTC register.
```
Sorry, I don't understand the need of this waiting. Why does it matter if the value is written to the RTC register (I assume you mean the one in LFCLK domain) or not at this point?

pabigot commented:
> Both type of lateness can be discovered in the callback by comparing `ticks` which are part of the callback with `counter_read()` performed in the callback context. Wouldn't that satisfy your requirement while keeping api simpler?

No.  Tests in the callback can only identify late-to-callback; you can't distinguish it from late-to-set.  Also in this solution if the alarm action is connected to a hardware-triggered task (PPI) there may be no callback to use, or the callback would be invoked without the associated task having occurred, with no way to detect the issue.

pabigot commented:
This solution is making what I see as an API change (behavior when an alarm is set too late, #12626) by implementing in the Nordic RTC counter behavior that AFAIK is not specified in the counter API.  Am I wrong about that, or has this been discussed in the API telecon and gotten approval from other vendors who will be expected to make sure their own implementations conform?  @carlescufi @MaureenHelm

nordic-krch commented:
> This solution is making what I see as an API change (behavior when an alarm is set too late, #12626) by implementing in the Nordic RTC counter behavior that AFAIK is not specified in the counter API. Am I wrong about that, or has this been discussed in the API telecon and gotten approval from other vendors who will be expected to make sure their own implementations conform?

I haven't been discussed. In fact it would be good to specify behavior on late alarms in the api.

nordic-krch commented:
> No. Tests in the callback can only identify late-to-callback; you can't distinguish it from late-to-set. Also in this solution if the alarm action is connected to a hardware-triggered task (PPI) there may be no callback to use, or the callback would be invoked without the associated task having occurred, with no way to detect the issue.

@pabigot I don't think user should use PPI with counter API. It's vendor specific and better would be to use nrfx_rtc or custom rtc driver. That is because counter driver will not guaranty that there won't be more PPI events than scheduled (.e.g even after canceling alarm event will occur). Except for PPI what is the difference between late-to-set and late-to-callback from user perspective?

pabigot commented:
> @pabigot I don't think user should use PPI with counter API. It's vendor specific and better would be to use nrfx_rtc or custom rtc driver. That is because counter driver will not guaranty that there won't be more PPI events than scheduled (.e.g even after canceling alarm event will occur).

Which is why in #12625 I also specified that cancelling should indicate the state of the alarm.  That didn't go anywhere and is still a gap.

In any case, this comes down to a Zephyr architectural decision about whether to be as platform-independent as possible, even if that means blocking use of capabilities that are present on the platform I'm using.  I prefer infrastructure to be flexible and to allow me (as an application designer) to decide what features I want to use, even if the resulting application isn't portable.  I don't expect to win that battle, so if you really want to rule out the use of PPI with the Zephyr counter API I can't say much other than I think that should be a project-level decision, not one made incidental to a vendor-specific change.  Nordic isn't the only vendor with PPI-like functionality.

nordic-krch commented:
@pabigot I still don't see the benefit of distinction between late-to-set and late-to-callback. Regarding using counter and enabling RTC events through PPI: it's not that i want to limit it, I think it just don't make much sense because it's very hard (or impossible) to control appearance of events in generic way. In other words, you can do that if you control timings and avoid corner situations (like setting CC late).

But i agree that we may not limit API and give the user possibility to control how driver acts when late-to-set is detected. As you suggested we can add `flags` in `counter_alarm_cfg` (make `absolute` one of the flags).

anangl commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -81,18 +94,121 @@ static u32_t counter_nrfx_read(struct device *dev)
+		 */
+		LOG_INST_INF(nrfx_config->log,
+			"Delay introduced due to risk of pre-firing.");
+		NRFX_DELAY_US(LF_TICK_US);
```
What about using the TICK event instead?

anangl replied:
Let me answer myself, it won't work that well with a prescaler.

nordic-krch replied:
discussed offline with @anangl and fixed algorithm so that delay is no longer needed. Updated algorithm description.

nordic-krch commented:
I've added description of the algorithm (comment above `set_cc()`) and reworked the algorithm as i've found issues with previous one. 

@anangl your comments applied.

@pabigot, regarding late-to-set reaction (callback or error). Can we handle in another PR (API change)? Focus here on algorithm and proper detection of late setting case. Same with regards to DT vs Kconfig location. I will try to improve description of guard-period option and make the default one longer.

pabigot commented:
> @pabigot, regarding late-to-set reaction (callback or error). Can we handle in another PR (API change)?

I've updated [my review](https://github.com/zephyrproject-rtos/zephyr/pull/15510#pullrequestreview-237775232) with my current assessment and rationale for rejecting.

I don't like being a hard-case here, because I've been on the other end, but there's too much of Zephyr that's been designed from only one perspective leaving inconsistencies in cross-platform behavior and incomplete functionality.  I think my points are relevant and at least deserve evaluation from a wider set of reviewers before they are dismissed.

nordic-krch commented:
@pabigot I would like to keep focus of this PR on actual algorithm for detecting late-to-set  case. Current assumption in that PR is that we don't distinguish between late-to-set and late-to-callback. Further steps can involve adding flag to `counter_set_alarm` to get error on late-to-set (that's when `set_cc()` returns `-1`).

pabigot commented:
> @pabigot I would like to keep focus of this PR on actual algorithm for detecting late-to-set case.

I'm aware of that.  But as I understand this PR you are changing the existing behavior produced by the counter API in a late-to-set situation (1) only on Nordic RTC, and (2) without showing any evidence that there's agreement in the Zephyr community on what behavior in a late-to-set situation should be.

I'm not going to approve that.

If you want to this merged, find somebody else for your second technical review and get somebody to dismiss my review.  That's the way Zephyr's process currently works.

nordic-krch commented:
> I'm aware of that. But as I understand this PR you are changing the existing behavior produced by the counter API in a late-to-set situation (1) only on Nordic RTC, and (2) without showing any evidence that there's agreement in the Zephyr community on what behavior in a late-to-set situation should be.

Late-to-set case is not specified in counter API but I would rather assume that user expectation is that it will rather expire immediately or produce an error than trigger alarm after full turnaround occurs (may be in minutes or days).

I will work towards proposing counter API extension that clarifies it. As vendors will not support it (initially), it needs to be optional.

pabigot commented:
> Late-to-set case is not specified in counter API but I would rather assume that user expectation is that it will rather expire immediately or produce an error than trigger alarm after full turnaround occurs (may be in minutes or days).

Exactly: that's an assumption.  Waiting for the next lap is probably the existing behavior, and may be what people want or are unknowingly relying on.

> I will work towards proposing counter API extension that clarifies it. As vendors will not support it (initially), it needs to be optional.

Thank you.

nordic-krch committed:
tests: drivers: counter: Add test for late setting and short relative

Added tests which verifies that:
- driver will detect when absolute alarm is set too late.
- short relative alarm is supported
- canceled alarm is not triggering user callback

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
#16252 introduces flags for configuration of alarm. Once it goes in, it will be used by this patch to conditionally enable detection of late alarm setting.

nordic-krch committed:
drivers: counter: nrf_timer: Add detection of late alarm request

    Implemented latest extensions to the counter API related to
    detection of alarms being set too late and short relative alarms.
    Implementation could not be realized on nrfx_timer driver thus
    driver has been reimplemented based on nrf_timer hal.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@anangl @pizi-nordic I would like to finally get that PR in. As counter API PRs got in, scope of this PR has narrowed down to changes in 2 nordic drivers and counter test.

pizi-nordic commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -25,16 +25,17 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL);
-
-#define TOP_CH 0
-#define COUNTER_TOP_INT NRFX_RTC_INT_COMPARE0
+/* todo Change to use nrfx API when added (it's coming). */
```
How old is this line? ;-)

pizi-nordic commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -67,175 +68,248 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
+/* Function calculates distance between to values assuming that one first
+ * argument is in front and that values wrap.
+ */
+static u32_t ticks_sub(u32_t val, u32_t old, u32_t top, bool npow2)
```
Could we eliminate npow2 and just test the top value?

nordic-krch replied:
operations are faster when top is 2^n-1. Alternatively it could be a build time flag (`CONFIG_COUNTER_NRF_RTC_POW2_TOP`). When option is enabled (by default enabled) counter would return error if provided top is not 2^n-1. what do you think?

nordic-krch replied:
anyway, i will remove npow2 arguement. functions should get inlined anyway.

pizi-nordic commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -67,175 +68,248 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
 }
 
-static void disable(struct device *dev, u8_t id)
+static u32_t ticks_add(u32_t val1, u32_t val2, u32_t top, bool npow2)
```
Could we eliminate npow2?

nordic-krch replied:
done

pizi-nordic commented on drivers/counter/counter_nrfx_timer.c:
```diff
@@ -55,112 +65,171 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
+		/* if top is not 2^n-1 */
+		u32_t diff;
+
+		diff = (val > old) ? (val - old) : val + top + 1 - old;
```
Could we eliminate npow2?

nordic-krch replied:
done

nordic-krch commented:
@pizi-nordic your comments applied. @anangl ping

pizi-nordic commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -68,183 +63,246 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
-		cc_val -= (cc_val > get_dev_data(dev)->top) ?
-				get_dev_data(dev)->top : 0;
-	}
+	if (unlikely(top & (top + 1))) {
```
What about introducing local `static inline bool is_power_of_2(uint32_t value)`?

nordic-krch replied:
done

nordic-krch commented:
@anangl any chance to get review or approve?

nordic-krch commented:
@pabigot you've requested changes more than 4 months ego. Since that time a lot changed (mainly counter.h api got late detection support and driver adapted to that), can you take another look or withdraw your review?

anangl commented on drivers/counter/counter_nrfx_timer.c:
```diff
@@ -55,112 +65,173 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
+	}
+
+	/* if top is not 2^n-1 */
+	return (val > old) ? (val - old) : val + top + 1 - old;
```
It should be `(val >= old)`, otherwise `top + 1` can be returned.

nordic-krch replied:
fixed

anangl commented on drivers/counter/counter_nrfx_timer.c:
```diff
@@ -55,112 +65,173 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
+	nrf_timer_event_clear(reg, evt);
+
+	if (!absolute) {
+		trigger_late = val < (top / 2);
```
Could you provide some comment explaining the above? Why are the relative ticks compared against half of the top value? And why is this `bool` value used in the shifting right below?

nordic-krch replied:
added comment

anangl commented on drivers/counter/counter_nrfx_timer.c:
```diff
@@ -55,112 +65,173 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
+	return err;
 }
 
 static void disable(struct device *dev, u8_t id)
```
Why separate function, when it is called only by `cancel_alarm` that differs solely by returning `0`?

nordic-krch replied:
merged to `cancel_alarm`

anangl reviewed (commented):
A first bunch of comments, I haven't finished reviewing yet.

nordic-krch commented on drivers/counter/counter_nrfx_timer.c:
```diff
@@ -173,121 +241,155 @@ static int counter_nrfx_set_top_value(struct device *dev,
 			TIMER_PRESCALER_PRESCALER_Msk,			       \
 			"TIMER prescaler out of range");		       \
+	DEVICE_DECLARE(timer_##idx);					       \
+	void timer_##idx##_irq_handler(void)				       \
```
done (in both files)

nordic-krch commented on drivers/counter/counter_nrfx_timer.c:
```diff
@@ -17,17 +17,22 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL);
 
+/* todo Change to use nrfx API when added (it's coming). */
+#define CH2EVT(ch) \
+	(nrf_timer_event_t)(offsetof(NRF_TIMER_Type, EVENTS_COMPARE[ch]))
```
applied nrfx API

nordic-krch commented on drivers/counter/counter_nrfx_timer.c:
```diff
@@ -55,112 +65,173 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
 
-static u32_t counter_nrfx_read(struct device *dev)
+/* Return true if value equals 2^n - 1 */
+static inline bool is_power_of_2(u32_t val)
```
renamed

nordic-krch commented on drivers/counter/counter_nrfx_timer.c:
```diff
@@ -55,112 +65,173 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
+	if (likely(is_power_of_2(top))) {
+		sum = sum & top;
+	} else {
+		sum = sum > top ? sum - top : sum;
```
right, it was copied from rtc where 24 bit limit ensures that there is no overflow. Fixed it for timer case.

nordic-krch commented on drivers/counter/counter_nrfx_timer.c:
```diff
@@ -173,121 +244,155 @@ static int counter_nrfx_set_top_value(struct device *dev,
 
-	return 0;
+	for (u32_t i = 0; i < counter_get_num_of_channels(dev); i++) {
+		alarm_irq_handle(dev, ID_TO_CC(i));
```
applied suggestion

anangl commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -257,38 +324,124 @@ static int ppi_setup(struct device *dev)
+		if (err == 0) {
+			nrf_rtc_cc_set(rtc, top_ch, cfg->ticks);
+			/* mark channel as busy */
+			nrfx_config->ch_data[top_ch].callback = (void *)1;
```
I'm afraid this is not a good idea. If one channel is needed for handling the top value, it should not be made available for API users. With the above implementation, the driver reports via `counter_get_num_of_channels` that certain number of channels is available for use, but after the TOP value is set to something different than what is reported by `counter_get_max_top_value`, apparently one of these channels becomes unusable (permanently busy).

nordic-krch replied:
on the other hand, it is very likely that user will use default top value and in that case it's a waste. Alternatively, i can add compile time option per instance to enable custom top value or not. what do you think?

nordic-krch replied:
I agree that it may be confusing but as i mentioned we might need this one more channel badly when counter will be used as system timer and we would like to share it with bluetooth to free up one RTC. I think that best would be to add `fixed_top` flag to instance configuration and do it in similar way as `ppi_wrap` so that if all instances are using fixed top value then support for custom top is removed. That can also simplify functions for ticks operations (add, sub). However, i started this and it will be more changes so would propose to do it as another PR and get that one in. Do you agree?

nordic-krch replied:
@anangl I've created issue for that #19569.

anangl replied:
The idea of `fixed_top` flag seems to be reasonable but I'm not convinced that this is the right way to proceed: to introduce this confusing allocation of one channel in this PR and later, potentially, remove it with another one, calling it an enhancement.

nordic-krch replied:
ok, i've done following: added fixed_top flag per instance. Also added a kconfig flag which is set if all instances have fixed_top set. If yes much of the counter code is removed (significant driver code size reduction 1,5k -> 1k). Added dedicated test suite to validate correct behavior when fixed_top is set for all instances.

Additionally, enabled ppi-wrap on one instance in the test (it revealed that ppi-wrap was broken) added ppi-wrap fix.

pabigot reviewed (commented):
My causes for rejection in https://github.com/zephyrproject-rtos/zephyr/pull/15510#pullrequestreview-237775232 appear to be obsolete: the API change has gotten past any guardians, and as far as I can tell the "guard period" is no longer a millisecond offset specified in devicetree.

My reasons for not approving stand: this solution (and the already-merged API) is not based on a [well-defined concept of what a counter is and how it should behave](https://github.com/zephyrproject-rtos/zephyr/pull/16252#issuecomment-494174959).  The "guard period" concept is not necessary to achieve a solution to #12626.  I don't know if the needs of #12626 are satisfied by the already-merged API changes, but this PR would have no effect on that.

So: rejection withdrawn, approval not forthcoming.

nordic-krch commented on drivers/counter/counter_nrfx_timer.c:
```diff
@@ -55,112 +61,167 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
+
+static void set_cc_int_pending(struct device *dev, u8_t chan)
+{
+	get_dev_data(dev)->cc_int_pending |= BIT(chan);
```
`set_cc_int_pending` is called when RTC interrupts are temporarily disabled.

anangl replied:
Really? I pointed out the implementation for TIMER but even for RTC I can't see this temporary disabling of interrupts.

nordic-krch commented on drivers/counter/counter_nrfx_timer.c:
```diff
@@ -55,112 +65,173 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
+	u32_t diff;
+	u32_t max_abs_val;
+
+	max_abs_val = top - get_dev_data(dev)->guard_period;
```
added assert to setting guard period

nordic-krch committed:
drivers: counter: nrf_rtc: Fix wrapping with PPI

PPI allocation and freeing was not handled correctly. Additionally,
RTC event was not enabled when PPI was enabled which resulted in
lack of RTC counter clearing.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: counter: Enable ppi-wrap feature on one instance

Enable ppi-wrap feature to test it on at least one instance.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: counter: Add test suite for nrf rtc fixed top value feature

Added test suite which performs sanity check on instances with
fixed-top feature enabled.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented on drivers/counter/counter_nrfx_timer.c:
```diff
@@ -173,121 +234,153 @@ static int counter_nrfx_set_top_value(struct device *dev,
 {
-	const struct counter_nrfx_config *nrfx_config = get_nrfx_config(dev);
-	const nrfx_timer_t *timer = &nrfx_config->timer;
+	__ASSERT_NO_MSG(guard < get_top_value(dev));
```
added assertion in `set_cc`

anangl replied:
Yes, in `counter_nrfx_rtc.c`, but not in `counter_nrfx_timer.c`.

nordic-krch commented:
@anangl 
> Really? I pointed out the implementation for TIMER but even for RTC I can't see this temporary disabling of interrupts.

Interrupts are disabled whenever alarm expires (in interrupt) or alarm is canceled and `set_cc` is only called when channel is free (expired or never used) so on entering `set_cc` interrupt is disabled.

I've added assert checking that interrupt is disabled on entering `set_cc` in timer and rtc implementation.

anangl commented:
> @anangl
> 
> > Really? I pointed out the implementation for TIMER but even for RTC I can't see this temporary disabling of interrupts.
> 
> Interrupts are disabled whenever alarm expires (in interrupt) or alarm is canceled and `set_cc` is only called when channel is free (expired or never used) so on entering `set_cc` interrupt is disabled.

What is disabled is only the corresponding event as the interrupt source. The RTC or TIMER interrupt itself is still enabled and the interrupt handler can still be executed on the top compare event or some other channel compare event. And that was my point. In certain circumstances (let's say, `set_channel_alarm` called for different channels from different contexts, one from a higher priority interrupt and the other one from an alarm handler), when `get_dev_data(dev)->cc_int_pending |= BIT(chan);` is interrupted, you can end up in executing some alarm handler twice or not at all. I know that this is quite rare scenario, seems rather contrived, but it is not impossible to fall into, and when such thing occurs, it may be very hard to guess what actually happened.

anangl commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -68,226 +74,428 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
 {
-	nrfx_rtc_disable(&get_nrfx_config(dev)->rtc);
+	nrf_rtc_task_trigger(get_nrfx_config(dev)->rtc, NRF_RTC_TASK_STOP);
+	nrf_rtc_task_trigger(get_nrfx_config(dev)->rtc, NRF_RTC_TASK_CLEAR);
```
Why `CLEAR`? The counter API does not say that the counter should be cleared after it is stopped.

nordic-krch replied:
on the other hand it does not say that it shouldn't 😄. If we assume that counter is not cleared after stop and not cleared when started then counter API is missing `counter_clear` call. On the other hand for some implementations it might be hard to maintain  "no cleaning" approach. For example in NRF_TIMER if `TASKS_SHUTDOWN` is used to stop then counter is reset.

I think that it would be better to add `counter_pause` (some may return not supported) in a future and specify that `stop` clears the counter.

anangl replied:
My comment was not towards preserving the current value while the counter is stopped, as the API does not specify that counters should provide that. The thing is that everything should work without this clearing. And it was, with the previous implementation using the `nrfx_rtc` driver. Thus, if this clearing is needed now, it may indicate that something (a test?) is relying on a non-specified behavior.

nordic-krch replied:
removed clearing. test is passing.

anangl commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -68,226 +74,428 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
+	if (likely(is_bit_mask(top))) {
+		sum = sum & top;
+	} else {
+		sum = sum > top ? sum - top : sum;
```
Shouldn't it be `sum - (top + 1)` when `sum > top`?

nordic-krch replied:
done

anangl commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -68,226 +74,428 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
+				get_nrfx_config(dev)->ch_data[chan].callback =
+									NULL;
+			}
+		} else if (diff == 0) {
```
This case was already handled in `if (diff == 1)` block [above](https://github.com/zephyrproject-rtos/zephyr/pull/15510/files#diff-cdca186ce3c3dea4b9e41fdc13ff7f2eR243), wasn't it?

nordic-krch replied:
it's a new diff calculated after setting CC (now is read after setting CC). It maybe that procedure was interrupted and what was 5 ticks in the future, now is 1 tick in the future. We don't know when interruption occurred. It may occurred before setting CC and it was set for COUNTER+1 and CC event will not occur.

anangl replied:
Right, I missed that. This implementation is not really reviewer-friendly. :wink:
A word of comment on this re-reading of `now`, explaining why it is done, would help a bit probably.

nordic-krch replied:
added comment, I've also extracted `CC == COUNTER+1` case handling to a function and added description of that function.

anangl commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -68,226 +74,428 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
+	if (diff == 1) {
+		val = ticks_add(dev, val, 1, top);
+		nrf_rtc_cc_set(rtc, chan, val);
+		get_dev_data(dev)->cc_adjusted |= BIT(chan);
```
`cc_adjusted` should be also protected with atomic access, just like `cc_int_pending`.

nordic-krch replied:
done

nordic-krch commented:
@anangl can you recheck?

anangl commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -68,226 +78,456 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
-	nrfx_rtc_cc_set(rtc, TOP_CH, cfg->ticks, false);
+	diff = ticks_sub(dev, val, now, top);
+	if (diff == 1) {
+		/* If CC cannot be set to COUNTER+1 because that will not
```
This initial "If" seems to be unwanted.

nordic-krch replied:
sure, done

anangl commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -25,17 +25,24 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL);
 	u32_t top;
+	u32_t guard_period;
+	/* Store channel interrupt pending and CC adjusted flags. */
+	atomic_t ipend_adj;
```
Nice and creative name, with a bit of a mystery.
No need to change it, I just wanted to express my admiration. :smile:

anangl reviewed (approved):
Looks good. I left only some minor comments.

nordic-krch commented on drivers/counter/counter_nrfx_timer.c:
```diff
@@ -55,112 +62,186 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
+static int set_cc(struct device *dev, u8_t id, u32_t val, u32_t flags)
+{
+	__ASSERT_NO_MSG(get_dev_data(dev)->guard_period <
+			nrf_timer_cc_read(get_nrfx_config(dev)->timer, TOP_CH));
```
done

nordic-krch commented on drivers/counter/counter_nrfx_timer.c:
```diff
@@ -55,112 +62,186 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
+	NRF_TIMER_Type  *reg = get_nrfx_config(dev)->timer;
+	u8_t chan = ID_TO_CC(id);
+	nrf_timer_event_t evt = nrf_timer_compare_event_get(chan);
+	u32_t top = counter_get_top_value(dev);
```
done

nordic-krch commented on drivers/counter/counter_nrfx_rtc.c:
```diff
@@ -68,226 +78,456 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
-	for (int i = 0; i < counter_get_num_of_channels(dev); i++) {
-		/* Overflow can be changed only when all alarms are
-		 * disables.
+	/* Special handling required if value is next tick as that will not
```
done

nordic-krch commented:
@anangl thanks for review.

nordic-krch commented:
@carlescufi can you merge it?
